### ISSUE TITLE
a way for selective interpretation of `struct` definition ?

### ISSUE THREAD

**Issue by aviatesk** [JuliaDebug/LoweredCodeUtils.jl#47]
I would like to use this package for selective interpretation of definitions (i.e. method/struct/macro, etc), and now I'm facing this issue:

```julia
using LoweredCodeUtils, JuliaInterpreter

src = (Meta.@lower begin
    # I want to interpret this definition
    struct Foo
        b::Bool
        Foo(b) = new(b)
    end

    # I don't want to interpret this call site
    Foo(false)
end).args |> first

lines = src.code
edges = CodeEdges(src)

isdef = lines_required(:Foo, src, edges)

frame = Frame(@__MODULE__, src)
selective_eval_fromstart!(frame, isdef, true)
```
```
ERROR: circular type parameter constraint in definition of Foo
Stacktrace:
 [1] is_vararg_type at /Users/aviatesk/julia/packages/JuliaInterpreter/src/utils.jl:252 [inlined]
 [2] _any(::typeof(JuliaInterpreter.is_vararg_type), ::Array{Any,1}, ::Colon) at ./reduce.jl:781
 [3] #any#636 at ./reducedim.jl:732 [inlined]
 [4] any at ./reducedim.jl:732 [inlined]
 [5] prepare_call(::Any, ::Array{Any,1}; enter_generated::Bool) at /Users/aviatesk/julia/packages/JuliaInterpreter/src/construct.jl:251
 [6] get_call_framecode(::Array{Any,1}, ::JuliaInterpreter.FrameCode, ::Int64; enter_generated::Bool) at /Users/aviatesk/julia/packages/JuliaInterpreter/src/localmethtable.jl:63
 [7] evaluate_call_recurse!(::Any, ::Frame, ::Expr; enter_generated::Bool) at /Users/aviatesk/julia/packages/JuliaInterpreter/src/interpret.jl:225
 [8] evaluate_call_recurse! at /Users/aviatesk/julia/packages/JuliaInterpreter/src/interpret.jl:202 [inlined]
 [9] eval_rhs(::Any, ::Frame, ::Expr) at /Users/aviatesk/julia/packages/JuliaInterpreter/src/interpret.jl:389
 [10] step_expr!(::Any, ::Frame, ::Any, ::Bool) at /Users/aviatesk/julia/packages/JuliaInterpreter/src/interpret.jl:527
 [11] step_expr! at /Users/aviatesk/julia/packages/JuliaInterpreter/src/interpret.jl:582 [inlined]
 [12] selective_eval!(::Any, ::Frame, ::BitArray{1}, ::Bool) at /Users/aviatesk/julia/packages/LoweredCodeUtils/src/codeedges.jl:794
 [13] selective_eval_fromstart! at /Users/aviatesk/julia/packages/LoweredCodeUtils/src/codeedges.jl:819 [inlined]
 [14] selective_eval_fromstart!(::Frame, ::BitArray{1}, ::Bool) at /Users/aviatesk/julia/packages/LoweredCodeUtils/src/codeedges.jl:822
 [15] top-level scope at REPL[13]:20
```
(fwiw, the code snippet above will yield segmentation fault on Julia@master)

For this case, the selected statements are below:
```julia
julia> println.(zip(isdef, src.code));
(false, :(global Foo))
(false, :(const Foo))
(true, :(Core.svec()))
(true, :(Core.svec(:b)))
(true, :(_1 = Core._structtype(Main, :Foo, %3, %4, false, 1)))
(false, :(Core._setsuper!(_1, Core.Any)))
(true, :($(Expr(:isdefined, :Foo))))
(true, :(unless %7 goto %16))
(true, :Foo)
(true, :(Core._equiv_typedef(%9, _1)))
(true, :(unless %10 goto %14))
(true, :(_1 = %9))
(true, :(goto %15))
(true, :(Foo = _1))
(true, :(goto %17))
(true, :(Foo = _1))
(false, :(_1))
(false, :(Core.svec(Bool)))
(false, :(Core._typebody!(%17, %18)))
(false, :(global Foo))
(true, :($(Expr(:method, :Foo))))
(true, :(Core.Typeof(Foo)))
(true, :(Core.svec(%22, Core.Any)))
(true, :(Core.svec()))
(true, :(Core.svec(%23, %24, $(QuoteNode(:(#= REPL[13]:7 =#))))))
(true, :($(Expr(:method, :Foo, :(%25), CodeInfo(quote
    Foo
    Core.fieldtype(%1, 1)
    Base.convert(%2, b)
    %new(%1, %3)
    return %4
end)))))
(false, :(Foo(false)))
(false, :(return %27))
```
Obviously some statements for first time definitions for this `Foo` are not selected by `line_required`, and that seems to cause this error.

fyi, If I made required lines by
```
isdef = LoweredCodeUtils.ismethod.(lines) # find a :method def bound to `Foo`
lines_required!(isdef, src, edges)          # fill up `isdef` with required statements
```
it still yields same result.

---

So my question is, is this expected and if so, could you tell me "valid" way for selective interpretation of struct definitions ? I know the same approach works for usual method definitions, but maybe LoweredCodeUtils won't support selective interpretation of `struct` definitions ? (just guessing because Revise can't update struct definitions anyway)

**Comment by timholy** [comment]
Nice work, you've discovered the innermost secrets of the universe :laughing:.

Yes, the `lines_required` is not getting everything it needs, somehow. This is important to fix. I'm pretty busy the rest of the day, so what I'd recommend is take a look at https://github.com/timholy/Revise.jl/blob/ed86caf56bd29445fffdf44fc2689248e9531a2f/src/lowered.jl#L50-L122 and see if it helps.

We might want to transfer some of that logic here, it's not great to have it split this way.

**Comment by aviatesk** [comment]
hehe, thanks for your kind pointer :) I will try Revise's code first (and then may try to upstream some code into LoweredCodeUtils if it seems appropriate)

**Comment by aviatesk** [comment]
okay, here is what I've tracked so far:

I tried this diff:
```diff
diff --git a/src/lowered.jl b/src/lowered.jl
index e754c6b..0ae96da 100644
--- a/src/lowered.jl
+++ b/src/lowered.jl
@@ -94,6 +94,24 @@ function minimal_evaluation!(predicate, methodinfo, src::Core.CodeInfo, mode::Sy
             if isa(lhs, Symbol)
                 isrequired[edges.byname[lhs].succs] .= true  # mark any `const` statements or other "uses" in this block
             end
+        elseif mode === :def
+            defs = Set(GlobalRef(Core, sym) for sym in (:_abstracttype,
+                                                        :_structtype,
+                                                        :_primitivetype,
+                                                        :_setsuper!,
+                                                        :_equiv_typedef,
+                                                        :_typebody!,
+                                                        ))
+            if isexpr(stmt, :(=))
+                stmt = stmt.args[2] # rhs
+            end
+            if isexpr(stmt, :call)
+                if stmt.args[1] in defs
+                    isrequired[i] = true
+                end
+            elseif isexpr(stmt, :global) || isexpr(stmt, :const)
+                isrequired[i] = true
+            end
         end
     end
     # Check for docstrings
@@ -104,9 +122,11 @@ function minimal_evaluation!(predicate, methodinfo, src::Core.CodeInfo, mode::Sy
         end
     end
     # All tracked expressions are marked. Now add their dependencies.
-    # LoweredCodeUtils.print_with_code(stdout, src, isrequired)
+    @info "before lines_required!"
+    LoweredCodeUtils.print_with_code(stdout, src, isrequired)
     lines_required!(isrequired, src, edges; exclude_named_typedefs=mode===:sigs)
-    # LoweredCodeUtils.print_with_code(stdout, src, isrequired)
+    @info "after lines_required!"
+    LoweredCodeUtils.print_with_code(stdout, src, isrequired)
     add_dependencies!(methodinfo, edges, src, isrequired)
     return isrequired, evalassign
 end
```

this is a bit ugly, but now we've marked necessary "definition" calls (before `lines_required!`):
```julia
julia> lwr = @lower begin
           # I want to interpret these definitions
           abstract type Bar end
           struct Bar1 <: Bar
               bar
           end
       
           # I don't want to interpret statements below
           r = rand(1000000000000)
           bar = Bar1(r)
           show(bar)
       end;

julia> src = first(lwr.args);

julia> isrq, _ = Revise.minimal_evaluation!(Revise.MethodInfo(), src, :def)
[ Info: before lines_required!
No IR statement printer available on this version of Julia, just aligning statements.
 1 t global Bar
 2 t const Bar
 3 f Core.svec()
 4 t _1 = Core._abstracttype(Main, :Bar, %3)
 5 t Core._setsuper!(_1, Core.Any)
 6 t Core._typebody!(_1)
 7 f $(Expr(:isdefined, :Bar))
 8 f unless %7 goto %11
 9 t _3 = Core._equiv_typedef(Bar, _1)
10 f goto %12
11 f _3 = false
12 f unless _3 goto %14
13 f goto %15
14 f Bar = _1
15 t global Bar1
16 t const Bar1
17 f Core.svec()
18 f Core.svec(:bar)
19 t _2 = Core._structtype(Main, :Bar1, %17, %18, false, 1)
20 t Core._setsuper!(_2, Bar)
21 f $(Expr(:isdefined, :Bar1))
22 f unless %21 goto %30
23 f Bar1
24 t Core._equiv_typedef(%23, _2)
25 f unless %24 goto %28
26 f _2 = %23
27 f goto %29
28 f Bar1 = _2
29 f goto %31
30 f Bar1 = _2
31 f _2
32 f Core.svec(Core.Any)
33 t Core._typebody!(%31, %32)
34 t global Bar1
35 t $(Expr(:method, :Bar1))
36 f Core.Typeof(Bar1)
37 f Core.svec(%36, Core.Any)
38 f Core.svec()
39 f Core.svec(%37, %38, $(QuoteNode(:(#= none:5 =#))))
40 t $(Expr(:method, :Bar1, :(%39), CodeInfo(quote
    Bar1
    %new(%1, bar)
    return %2
end)))
41 f rand(1000000000000)
42 f r = %41
43 f Bar1(r)
44 f bar = %43
45 f show(bar)
46 f return %45
```

but now `line_required!` doesn't work as expected; it marks the call sites after the definitions, which I wouldn't like to select:
```julia
# this output follows the above case
[ Info: after lines_required!
No IR statement printer available on this version of Julia, just aligning statements.
 1 t global Bar
 2 t const Bar
 3 t Core.svec()
 4 t _1 = Core._abstracttype(Main, :Bar, %3)
 5 t Core._setsuper!(_1, Core.Any)
 6 t Core._typebody!(_1)
 7 t $(Expr(:isdefined, :Bar))
 8 t unless %7 goto %11
 9 t _3 = Core._equiv_typedef(Bar, _1)
10 t goto %12
11 t _3 = false
12 t unless _3 goto %14
13 t goto %15
14 t Bar = _1
15 t global Bar1
16 t const Bar1
17 t Core.svec()
18 t Core.svec(:bar)
19 t _2 = Core._structtype(Main, :Bar1, %17, %18, false, 1)
20 t Core._setsuper!(_2, Bar)
21 t $(Expr(:isdefined, :Bar1))
22 t unless %21 goto %30
23 t Bar1
24 t Core._equiv_typedef(%23, _2)
25 t unless %24 goto %28
26 t _2 = %23
27 t goto %29
28 t Bar1 = _2
29 t goto %31
30 t Bar1 = _2
31 t _2
32 t Core.svec(Core.Any)
33 t Core._typebody!(%31, %32)
34 t global Bar1
35 t $(Expr(:method, :Bar1))
36 t Core.Typeof(Bar1)
37 t Core.svec(%36, Core.Any)
38 t Core.svec()
39 t Core.svec(%37, %38, $(QuoteNode(:(#= none:5 =#))))
40 t $(Expr(:method, :Bar1, :(%39), CodeInfo(quote
    Bar1
    %new(%1, bar)
    return %2
end)))
41 t rand(1000000000000)
42 t r = %41
43 t Bar1(r)
44 t bar = %43
45 f show(bar)
46 f return %45
```

... and now I feel a bit afraid to inspect into `lines_required!` 😅

**Comment by timholy** [comment]
Nice work! This is a good application for DebuggingUtilities `@showln`, so you can monitor `isrequired` at each "stage" of `lines_required!`. You can use it to discover that it's being added in the "named objects" phase:
```diff
diff --git a/src/codeedges.jl b/src/codeedges.jl
index 98d8e24..0941d1f 100644
--- a/src/codeedges.jl
+++ b/src/codeedges.jl
@@ -594,6 +594,7 @@ function lines_required!(isrequired::AbstractVector{Bool}, src::CodeInfo, edges:
     return lines_required!(isrequired, objs, src, edges; kwargs...)
 end
 
+using DebuggingUtilities
 function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo, edges::CodeEdges; exclude_named_typedefs::Bool=false)
     # Do a traveral of "numbered" predecessors
     # We'll mostly use generic graph traversal to discover all the lines we need,
@@ -632,6 +633,7 @@ function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo,
         add_obj!(isrequired, objsnew, obj, edges, norequire)
     end
     objs = objsnew
+    @showln isrequired
 
     # Compute basic blocks, which we'll use to make sure we mark necessary control-flow
     bbs = Core.Compiler.compute_basic_blocks(src.code)  # needed for control-flow analysis
@@ -647,6 +649,8 @@ function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo,
                 changed |= add_preds!(isrequired, idx, edges, norequire)
             end
         end
+        @showln isrequired
+
         # Handle named dependencies
         for (obj, uses) in edges.byname
             obj ∈ objs && continue
@@ -654,6 +658,7 @@ function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo,
                 changed |= add_obj!(isrequired, objs, obj, edges, norequire)
             end
         end
+        @showln isrequired
         # Add control-flow. For any basic block with an evaluated statement inside it,
         # check to see if the block has any successors, and if so mark that block's exit statement.
         # Likewise, any preceding blocks should have *their* exit statement marked.
@@ -684,6 +689,7 @@ function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo,
                 end
             end
         end
+        @showln isrequired
         # So far, everything is generic graph traversal. Now we add some domain-specific information.
         # New struct definitions, including their constructors, get spread out over many
         # statements. If we're evaluating any of them, it's important to evaluate *all* of them.
@@ -729,6 +735,7 @@ function lines_required!(isrequired::AbstractVector{Bool}, objs, src::CodeInfo,
             end
             idx += 1
         end
+        @showln isrequired
         iter += 1  # just for diagnostics
     end
     return isrequired
```
with the result
```julia
isrequired = Bool[0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
(in /home/tim/.julia/dev/LoweredCodeUtils/src/codeedges.jl:636)
isrequired = Bool[0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0]
(in /home/tim/.julia/dev/LoweredCodeUtils/src/codeedges.jl:652)
```
So it seems to be being added in the "SSA predecessors" phase. That's pretty crazy!

Taking a look at the `edges`, it seems pretty clear where it's gone wrong:
```julia
julia> edges = CodeEdges(src)
CodeEdges:
  Bar: assigned on [12], depends on [4], and used by [1, 2, 4, 5, 6, 7, 9, 18]
  Bar1: assigned on [26, 28, 33, 38], depends on [17, 24, 26, 28, 33, 37, 38], and used by [13, 14, 17, 18, 19, 21, 22, 24, 26, 28, 29, 32, 33, 34, 38]
  bar: assigned on [42], depends on [41], and used by [16, 43]
  Core.Any: assigned on ∅, depends on ∅, and used by [5, 30, 35]
  r: assigned on [40], depends on [39], and used by [41]
  statement  1 depends on [12] and is used by ∅
  statement  2 depends on [12] and is used by ∅
  statement  3 depends on ∅ and is used by [4, 5, 6, 9]
  statement  4 depends on [3, 12] and is used by [5, 6, 9, 12]
  statement  5 depends on [3, 4] and is used by ∅
  statement  6 depends on [3, 4] and is used by ∅
  statement  7 depends on [12] and is used by [8]
  statement  8 depends on [7] and is used by ∅
  statement  9 depends on [3, 4, 12] and is used by [10]
  statement 10 depends on [9] and is used by ∅
  statement 11 depends on ∅ and is used by ∅
  statement 12 depends on [3, 4] and is used by [1, 2, 4, 5, 6, 7, 9, 18]
  statement 13 depends on [26, 28, 33, 38] and is used by ∅
  statement 14 depends on [26, 28, 33, 38] and is used by ∅
  statement 15 depends on ∅ and is used by [17, 18, 22, 24, 29]
  statement 16 depends on [42] and is used by [17, 18, 22, 24, 29]
  statement 17 depends on [15, 16, 21, 24, 26, 28, 33, 38] and is used by [18, 22, 24, 26, 28, 29, 33, 38]
  statement 18 depends on [12, 15, 16, 17, 21, 24] and is used by ∅
  statement 19 depends on [26, 28, 33, 38] and is used by [20]
  statement 20 depends on [19] and is used by ∅
  statement 21 depends on [26, 28, 33, 38] and is used by [17, 18, 22, 24, 29]
  statement 22 depends on [15, 16, 17, 21, 24] and is used by [23]
  statement 23 depends on [22] and is used by ∅
  statement 24 depends on [15, 16, 17, 21, 26, 28, 33, 38] and is used by [17, 18, 22, 26, 28, 29, 33, 38]
  statement 25 depends on ∅ and is used by ∅
  statement 26 depends on [15, 16, 17, 21, 24, 28, 33, 37, 38] and is used by [13, 14, 17, 18, 19, 21, 22, 24, 28, 29, 32, 33, 34, 38]
  statement 27 depends on ∅ and is used by ∅
  statement 28 depends on [15, 16, 17, 21, 24, 26, 33, 37, 38] and is used by [13, 14, 17, 18, 19, 21, 22, 24, 26, 29, 32, 33, 34, 38]
  statement 29 depends on [15, 16, 17, 21, 24] and is used by [31]
  statement 30 depends on ∅ and is used by [31]
  statement 31 depends on [29, 30] and is used by ∅
  statement 32 depends on [26, 28, 33, 38] and is used by ∅
  statement 33 depends on [26, 28, 38] and is used by ∅
  statement 34 depends on [26, 28, 33, 38] and is used by [35]
  statement 35 depends on [34] and is used by [37]
  statement 36 depends on ∅ and is used by [37]
  statement 37 depends on [35, 36] and is used by [26, 28, 33, 38]
  statement 38 depends on [26, 28, 33, 37] and is used by ∅
  statement 39 depends on ∅ and is used by [40]
  statement 40 depends on [39] and is used by [41]
  statement 41 depends on [40] and is used by [42]
  statement 42 depends on [41] and is used by [16, 43]
  statement 43 depends on [42] and is used by [44]
  statement 44 depends on [43] and is used by ∅
```
Statement 16 is
```julia
Core.svec(:bar)
```
and it's presumably because your variable has the same name as a field!

Wild...I am a bit busy with other things so can't take this farther now, but I wanted to share my progress in case you haven't arrived at the same place.

**Comment by timholy** [comment]
Obviously there's a lot going on here. I'm going to call the core issue the inappropriate marking of links, and say this is fixed by #48. (Easiest LoweredCodeUtils fix ever!)

But I really like your design in your Revise diff for marking the different components of `struct` assignment, and think we should move towards a world in which we have a number of "tagging" methods in this package, some of which are moved from Revise. Something along the lines of `require_structdef!(isrequired, src, structname::Symbol)` much like your Revise diff would be an awesome start on that.

**Comment by aviatesk** [comment]
Thanks for your quick response and fix for this !



> But I really like your design in your Revise diff for marking the different components of `struct` assignment, and think we should move towards a world in which we have a number of "tagging" methods in this package, some of which are moved from Revise. Something along the lines of `require_structdef!(isrequired, src, structname::Symbol)` much like your Revise diff would be an awesome start on that.

yeah I agree with the line; I may try to craft something tomorrow or so :)
